com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/steps/CheckForOperationsInProgressStep.java

```diff
@@ -0,0 +1,150 @@
+import com.sap.cloud.lm.sl.cf.core.cf.services.ServiceOperation;
+import com.sap.cloud.lm.sl.cf.core.cf.services.ServiceOperationState;
+import com.sap.cloud.lm.sl.cf.core.cf.services.ServiceOperationType;
+import com.sap.cloud.lm.sl.cf.process.message.Messages;
+import com.sap.cloud.lm.sl.common.util.JsonUtil;
+
+@Component("checkForOperationsInProgressStep")
+public class CheckForOperationsInProgressStep extends AsyncFlowableStep {
+
+    private static final String LAST_SERVICE_OPERATION = "last_operation";
+    private static final String SERVICE_OPERATION_TYPE = "type";
+    private static final String SERVICE_OPERATION_STATE = "state";
+    private static final String SERVICE_OPERATION_DESCRIPTION = "description";
+
+    @Inject
+    private ServiceGetter serviceInstanceGetter;
+    @Inject
+    private EventsGetter eventsGetter;
+
+    @Override
+    protected StepPhase executeAsyncStep(ExecutionWrapper execution) throws Exception {
+        CloudControllerClient controllerClient = execution.getControllerClient();
+        String spaceId = StepsUtil.getSpaceId(execution.getContext());
+        List<CloudServiceExtended> servicesToProcess = getServices(execution);
+        List<CloudServiceExtended> existingServices = getExistingServices(servicesToProcess, controllerClient);
+
+        if (existingServices.isEmpty()) {
+            return StepPhase.DONE;
+        }
+        Map<CloudServiceExtended, ServiceOperation> servicesInProgressState = getServicesOperationInProgress(existingServices,
+            controllerClient, spaceId);
+
+        if (servicesInProgressState.isEmpty()) {
+            return StepPhase.DONE;
+        }
+
+        Map<String, ServiceOperationType> servicesOperationTypes = getServicesOperationTypes(servicesInProgressState);
+        getStepLogger().info(Messages.POLLING_IN_PROGRESS_SERVICES);
+        getStepLogger().debug(Messages.SERVICES_IN_PROGRESS, JsonUtil.toJson(servicesOperationTypes, true));
+        StepsUtil.setTriggeredServiceOperations(execution.getContext(), servicesOperationTypes);
+        Map<String, CloudServiceExtended> servicesData = getServicesData(servicesInProgressState);
+        StepsUtil.setServicesData(execution.getContext(), servicesData);
+        return StepPhase.POLL;
+    }
+
+    protected List<CloudServiceExtended> getServices(ExecutionWrapper execution) {
+        List<CloudServiceExtended> services = new ArrayList<>(1);
+        services.add(StepsUtil.getServiceToProcess(execution.getContext()));
+        return services;
+    }
+
+    private List<CloudServiceExtended> getExistingServices(List<CloudServiceExtended> servicesToProcess,
+        CloudControllerClient controllerClient) {
+        return servicesToProcess.parallelStream()
+            .map(service -> controllerClient.getService(service.getName(), false))
+            .filter(Objects::nonNull)
+            .map(service -> ImmutableCloudServiceExtended.builder()
```

Comment:
Why the copying is required?

Reply:
Because StepsUtil.setServicesData() require List<CloudServiceExtended>